optimize global state if not used

breml opened the pull request:
@mna I quickly put together the idea of optimizing the global state handling by removing all the not necessary code, if the state change expression is not used by a generated parser, as suggested by you in https://github.com/mna/pigeon/pull/52#issuecomment-340593754.

What do you think?

mna commented:
@breml Looks very nice! Only thing I can think of is if a grammar looks up the state in a code block, but doesn't have any state code block. The `p.cur.state` field would not exist and it would fail to compile, which may be confusing to the user.

It is probably an error in the user's grammar anyway, but one scenario where that may happen is if the grammar is in progress, and the state code block was not yet added. But I don't think it's worth adding complexity in the parser to handle that, the documentation can make it clear that if there's no state code block, the `current.state` field does not exist.

breml commented:
@mna Thanks for your feedback. I changed the optimization of the global state such that the user of pigeon has to opt-in for this optimization. This is, the code for the global state handling is only removed if the global state expression is not used in the grammar and the user provides the `-optimize-parser` flag.
Additionally I updated the documentation (in `doc.go`).
